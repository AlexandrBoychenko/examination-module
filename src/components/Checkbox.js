import React from 'react';
import { setLocalStorage, returnCurrentValue } from '../helpers'

class Checkbox extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            checked: []
        };
        this.handleChangeCheckbox = this.handleChangeCheckbox.bind(this);
    }

    componentDidMount() {
        setLocalStorage(this.props.id, this.props.right);
        this.compareWithState();
    }

    handleChangeCheckbox(e) {
        let currentValues = this.state.checked.slice();
        let targetPos = currentValues.indexOf(e.target.value);
        if (~targetPos) {
            currentValues.splice(targetPos, 1);
        } else {
            currentValues.push(e.target.value);
        }
        this.setState({checked: currentValues});
        this.props.context.onAnswerChangeCheckbox(e.target.value, this.props.id, 'Checkbox');
    }

    renderItems() {
        return this.props.items.map((item) => {
            return this.renderInput(item)
        });
    }

    compareWithState() {
        let arrayForChecked = [];
        let pastValues = this.props.pastValues;
        this.props.items.forEach((item) => {
            if (pastValues && pastValues[this.props.id]) {
                pastValues[this.props.id].forEach((propsItem) => {
                    if(propsItem === item) {
                        arrayForChecked.push(item.toString());
                    }
                })
            }
        });
        this.setState({checked: arrayForChecked});
    }

    isItemChecked(item) {
        return ~this.state.checked.indexOf(item.toString())
    }

    renderInput(item) {
        this.className = 'checkbox-list';
        return (
            <li key={item}>
                <label className="container container-checkbox">{item}
                    <input type="checkbox" name="answer" value={item}
                       onChange={this.handleChangeCheckbox}
                       checked={returnCurrentValue(this.props.parentState.length, this.isItemChecked(item))}
                    />
                    <span className="check-mark-checkbox"></span>
                </label>
            </li>
        )
    }

    render() {
        return (
            <ul className={this.className}>
                {this.renderItems()}
            </ul>
        )
    }
}

export default Checkbox;
