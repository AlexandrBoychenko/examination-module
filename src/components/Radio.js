import React from 'react';
import { setLocalStorage } from '../helpers'

class Radio extends React.Component {
    constructor(props) {
        super(props);
        this.handleChangeRadio = this.handleChangeRadio.bind(this);
    }

    componentDidMount() {
        setLocalStorage(this.props.id, this.props.right);
    }

    handleChangeRadio(e) {
        this.props.context.onAnswerChangeRadio(e.target.value, this.props.id);
    }
    
    renderItems() {
        return this.props.items.map((item) => {
            return this.renderInput(item)
        });
    }

    renderInput(item) {
        this.className = 'radio-list';
        return (
            <li key={item} >
                <label className="container">{item}
                    <input
                        type="radio"
                        name={this.props.id}
                        value={item}
                        onChange={this.handleChangeRadio}
                    />
                    <span className="check-mark"></span>
                </label>
            </li>
        );            
    }

    render() {
        return (
            <ul className={this.className}>
                {this.renderItems()}
            </ul>
        )
    }
}

export default Radio;
