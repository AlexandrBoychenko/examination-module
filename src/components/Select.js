import React from 'react';
import { setLocalStorage, returnCurrentValue } from '../helpers'

class Select extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            selected: false
        };
        this.handleChangeSelect = this.handleChangeSelect.bind(this);
    }

    componentDidMount() {
        setLocalStorage(this.props.id, this.props.right);
        this.compareWithState();
    }

    handleChangeSelect(e) {
        this.setState({selected: e.target.value});
        this.props.context.onAnswerChangeSelect(e.target.value, this.props.id, 'Select')
    }

    renderItems() {
        return this.props.items.map((item) => {
            return <option
                key={item}
                value={item}
            >{item}</option>
        });
    }

    compareWithState() {
        let pastValues = this.props.pastValues;
        for (let key in pastValues) {
            if (key === this.props.id) {
                this.setState({selected: pastValues[key]});
            }
        }
    }

    render() {
        return(
            <div className="select-style">
                <select
                    value={returnCurrentValue(this.props.parentState.length, this.state.selected)}
                    onChange={this.handleChangeSelect}>
                    <option value=''>Выберите ответ</option>
                    {this.renderItems()}
                </select>
            </div>
        )
    }
}

export default Select;
