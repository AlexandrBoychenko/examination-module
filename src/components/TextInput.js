import React from 'react';
import { setLocalStorage, returnCurrentValue } from '../helpers'

class TextInput extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            value: ''
        };
        this.handleChangeTextInput = this.handleChangeTextInput.bind(this);
    }

    componentDidMount() {
        setLocalStorage(this.props.id, this.props.right.toLowerCase());
        this.compareWithState();
    }

    handleChangeTextInput(e) {
        this.setState({value: e.target.value});
        this.props.context.onAnswerChangeTextInput(e.target.value, this.props.id, 'TextInput')
    };

    compareWithState() {
        let pastValues = this.props.pastValues;
        for (let key in pastValues) {
            if (key === this.props.id) {
                this.setState({value: pastValues[key]});
            }
        }
    }

    render() {
        return(
            <input type="text"
                   name="answer"
                   placeholder="Введите ваш ответ здесь"
                   value={returnCurrentValue(this.props.parentState.hasOwnProperty(this.props.id), this.state.value)}
                   onChange={this.handleChangeTextInput}
            />
        )
    }
}

export default TextInput
