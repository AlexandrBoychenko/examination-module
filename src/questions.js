export default [
    {
        id: "answer01",
        value: 'Сколько планет в солнечной системе?',
        items: [7, 5, 9, 8],
        right: 8,
        type: "radio",
        component: "AnswerInput",
        context: this
    },
    {
        id: "answer02",
        value: 'Что характерно для кометы?',
        items: [
            "Имеет хвост",
            "Вращается вокруг солнца",
            "Состоит из газа и пыли",
            "Существует только во внутренней области солнечной системы",
            "Не имеет ядра"
        ],
        right: [
            "Имеет хвост",
            "Вращается вокруг солнца",
        ],
        type: "checkbox",
        component: "AnswerInput",
        context: this
    },
    {
        id: "answer03",
        value: 'Какая из планет солнечной системы имеет наибольший объем?',
        items: null,
        right: 'Юпитер',
        type: null,
        component: "AnswerTextInput",
        context: this
    },
    {
        id: "answer04",
        value: 'Как называется естественный спутник Земли?',
        items: [
            "Ганимед",
            "Луна",
            "Фобос",
            "Титан",
            "Европа"
        ],
        right: "Луна",
        type: null,
        component: "AnswerSelect",
        context: this
    },
    {
        id: "answer05",
        value: 'Достигнул ли к настоящему моменту космический аппарат Voyager-2, сконструированный "NASA",' +
        ' пределов солнечной системы?',
        items: ["Да", "Нет"],
        right: "Да",
        type: "radio",
        component: "AnswerInput",
        context: this
    }
];
