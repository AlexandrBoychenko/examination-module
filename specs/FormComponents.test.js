import React from 'react';
import Radio from '../src/Components/Radio';
import Checkbox from '../src/Components/Checkbox';
import TextInput from '../src/Components/TextInput';
import Select from '../src/Components/Select';
import renderer from 'react-test-renderer';
import questions from '../src/questions';

test('It should be render correctly', () => {
    const types = ['Radio', 'Checkbox', 'TextInput', 'Select'];
    types.forEach((questionTypes) => {
        const component = renderer.create(
            getComponentExpression(questionTypes)
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
});

const getComponentExpression = (questionType) => {
    let pastValues;
    let parentState = pastValues = {};
    let prevTypes = [];
    for (let i = 0; i <= questions.length; i++) {
        if (questions[i].type === questionType) {
            prevTypes.push(questions[i].type);
            return setProps({
                ...questions[i],
                pastValues,
                parentState
            })
        }
    }
};

const setProps = (props) => {
    switch(props.type) {
        case 'Radio': return <Radio {...props} />;
        case 'Checkbox': return <Checkbox {...props} />;
        case 'TextInput': return <TextInput {...props} />;
        case 'Select': return <Select {...props} />;
        default: return null
    }
};

