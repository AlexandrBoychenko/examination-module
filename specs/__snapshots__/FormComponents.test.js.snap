// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should be render correctly 1`] = `
<ul>
  <li>
    <label
      className="container"
    >
      7
      <input
        checked={false}
        name="question01"
        onChange={[Function]}
        type="radio"
        value={7}
      />
      <span
        className="check-mark"
      />
    </label>
  </li>
  <li>
    <label
      className="container"
    >
      5
      <input
        checked={false}
        name="question01"
        onChange={[Function]}
        type="radio"
        value={5}
      />
      <span
        className="check-mark"
      />
    </label>
  </li>
  <li>
    <label
      className="container"
    >
      9
      <input
        checked={false}
        name="question01"
        onChange={[Function]}
        type="radio"
        value={9}
      />
      <span
        className="check-mark"
      />
    </label>
  </li>
  <li>
    <label
      className="container"
    >
      8
      <input
        checked={false}
        name="question01"
        onChange={[Function]}
        type="radio"
        value={8}
      />
      <span
        className="check-mark"
      />
    </label>
  </li>
</ul>
`;

exports[`It should be render correctly 2`] = `
<ul>
  <li>
    <label
      className="container container-checkbox"
    >
      Имеет хвост
      <input
        checked=""
        name="answer"
        onChange={[Function]}
        type="checkbox"
        value="Имеет хвост"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
  <li>
    <label
      className="container container-checkbox"
    >
      Вращается вокруг солнца
      <input
        checked=""
        name="answer"
        onChange={[Function]}
        type="checkbox"
        value="Вращается вокруг солнца"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
  <li>
    <label
      className="container container-checkbox"
    >
      Состоит из газа и пыли
      <input
        checked=""
        name="answer"
        onChange={[Function]}
        type="checkbox"
        value="Состоит из газа и пыли"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
  <li>
    <label
      className="container container-checkbox"
    >
      Существует только во внутренней области солнечной системы
      <input
        checked=""
        name="answer"
        onChange={[Function]}
        type="checkbox"
        value="Существует только во внутренней области солнечной системы"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
  <li>
    <label
      className="container container-checkbox"
    >
      Не имеет ядра
      <input
        checked=""
        name="answer"
        onChange={[Function]}
        type="checkbox"
        value="Не имеет ядра"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
</ul>
`;

exports[`It should be render correctly 3`] = `
<input
  name="answer"
  onChange={[Function]}
  placeholder="Введите ваш ответ здесь"
  type="text"
  value=""
/>
`;

exports[`It should be render correctly 4`] = `
<div
  className="select-style"
>
  <select
    onChange={[Function]}
    value={false}
  >
    <option
      value=""
    >
      Выберите ответ
    </option>
    <option
      value="Ганимед"
    >
      Ганимед
    </option>
    <option
      value="Луна"
    >
      Луна
    </option>
    <option
      value="Фобос"
    >
      Фобос
    </option>
    <option
      value="Титан"
    >
      Титан
    </option>
    <option
      value="Европа"
    >
      Европа
    </option>
  </select>
</div>
`;
